server_test.go

```diff
@@ -1421,14 +1442,65 @@ func TestServerProcessPublishInvalidTopic(t *testing.T) {
 }
 
 func TestServerProcessPublishACLCheckDeny(t *testing.T) {
-	s := New(&Options{
-		Logger: logger,
-	})
-	s.Serve()
-	defer s.Close()
-	cl, _, _ := newTestClient()
-	err := s.processPublish(cl, *packets.TPacketData[packets.Publish].Get(packets.TPublishBasic).Packet)
-	require.NoError(t, err) // ACL check fails silently
+	tests := []struct {
+		name             string
+		protocolVersion  byte
+		pk               packets.Packet
+		expectErr        error
+		expectReponse    []byte
+		expectDisconnect bool
+	}{
+		{"v4_QOS0", 4, *packets.TPacketData[packets.Publish].Get(packets.TPublishBasic).Packet,
+			nil, nil, false},
+		{"v4_QOS1", 4, *packets.TPacketData[packets.Publish].Get(packets.TPublishQos1).Packet,
+			packets.ErrNotAuthorized, nil, true},
+		{"v4_QOS2", 4, *packets.TPacketData[packets.Publish].Get(packets.TPublishQos2).Packet,
+			packets.ErrNotAuthorized, nil, true},
+
+		{"v5_QOS0", 5, *packets.TPacketData[packets.Publish].Get(packets.TPublishBasicMqtt5).Packet,
+			nil, nil, false},
+		{"v5_QOS1", 5, *packets.TPacketData[packets.Publish].Get(packets.TPublishQos1Mqtt5).Packet,
+			nil, packets.TPacketData[packets.Puback].Get(packets.TPubrecMqtt5NotAuthorized).RawBytes, false},
+		{"v5_QOS2", 5, *packets.TPacketData[packets.Publish].Get(packets.TPublishQos2Mqtt5).Packet,
+			nil, packets.TPacketData[packets.Pubrec].Get(packets.TPubrecMqtt5NotAuthorized).RawBytes, false},
+	}
+
+	for _, tt := range tests {
+		tt := tt
```

Comment:
@thedevop curious to the reason for this line

Reply:
This is to address this issue: https://github.com/golang/go/wiki/CommonMistakes.

Although as it is, tt := tt is not needed because these tests are not running in parallel, but out of habit I add that as often I do run test cases in parallel (only one line change).

Reply:
Very interesting, thank you!